src/compute-client/src/service.rs

```diff
@@ -296,30 +300,49 @@ where
                         entry.0.update_iter(lower.iter().map(|t| (t.clone(), -1)));
                         entry.0.update_iter(upper.iter().map(|t| (t.clone(), 1)));
-                        entry.1.append(&mut updates);
                         let new_frontier = entry.0.frontier().to_owned();
-                        if old_frontier != new_frontier {
-                            consolidate_updates(&mut entry.1);
-                            let mut ship = Vec::new();
-                            let mut keep = Vec::new();
-                            for (time, data, diff) in entry.1.drain(..) {
-                                if new_frontier.less_equal(&time) {
-                                    keep.push((time, data, diff));
+                        match (&mut entry.1, &mut updates) {
+                            (Err(_), _) => {
+                                // Subscribe is borked; nothing to do.
+                                None
+                            }
+                            (_, Err(text)) => {
+                                entry.1 = Err(text.clone());
+                                Some(Ok(ComputeResponse::SubscribeResponse(
+                                    id,
+                                    SubscribeResponse::Batch(SubscribeBatch {
+                                        lower: old_frontier,
+                                        upper: new_frontier,
+                                        updates: Err(text.to_string()),
+                                    }),
+                                )))
+                            }
+                            (Ok(stashed_updates), Ok(updates)) => {
+                                stashed_updates.append(updates);
+                                if old_frontier != new_frontier {
+                                    consolidate_updates(stashed_updates);
+                                    let mut ship = Vec::new();
+                                    let mut keep = Vec::new();
+                                    for (time, data, diff) in stashed_updates.drain(..) {
+                                        if new_frontier.less_equal(&time) {
+                                            keep.push((time, data, diff));
+                                        } else {
+                                            ship.push((time, data, diff));
+                                        }
+                                    }
+                                    entry.1 = Ok(keep);
+                                    Some(Ok(ComputeResponse::SubscribeResponse(
+                                        id,
+                                        SubscribeResponse::Batch(SubscribeBatch {
+                                            lower: old_frontier,
+                                            upper: new_frontier,
+                                            updates: Ok(ship),
```

Comment:
Is it feasible to do the byte count checking at `ship.push(..)` above, because that's actually where the bytes are allocated? That has the added benefit of protecting computed instances from oom'ing themselves too.

Reply:
I think they probably don't get this protection, because all of the contents that populate `ship` already had to be in memory (we are `drain(..)`ing the vector, so it isn't a new allocation at that point). Erroring before the `consolidate` call is potentially wrong, if it happens that we had 100MB that consolidate down to 1MB, and it will only reduce the memory use (if anything consolidates to a zero update).